### PR TITLE
JAVA-2337: invalidate the server if opening a connection to it throws any MongoSocketException

### PR DESCRIPTION
**Pull request by jyemin**
Jira issue: [JAVA-2337](https://jira.mongodb.org/browse/JAVA-2337)

### CODE REVIEW

File: driver-core/src/test/unit/com/mongodb/connection/DefaultServerSpecification.groovy [hunk 1]
```diff
@@ -97,8 +100,8 @@ class DefaultServerSpecification extends Specification {
 
         when:
         def latch = new CountDownLatch(1)
-        def receivedConnection
-        def receivedThrowable
+        def receivedConnection = null
```
Comment (jyemin) [inline driver-core/src/test/unit/com/mongodb/connection/DefaultServerSpecification.groovy]: IntelliJ warns that these variables may not have been initialized

File: driver-core/src/test/unit/com/mongodb/connection/DefaultServerSpecification.groovy [hunk 2]
```diff
@@ -146,36 +149,55 @@ class DefaultServerSpecification extends Specification {
...
         then:
-        thrown(MongoSecurityException)
+        def e = thrown(MongoException)
+        e.is(exceptionToThrow)
         1 * connectionPool.invalidate()
         1 * serverMonitor.connect()
+
+        where:
+        exceptionToThrow << [
+                new MongoSecurityException(createCredential('jeff', 'admin', '123'.toCharArray()), 'Auth failed'),
```
Comment (jyemin) [inline driver-core/src/test/unit/com/mongodb/connection/DefaultServerSpecification.groovy]: A bit of overkill here.

File: driver-core/src/test/unit/com/mongodb/connection/DefaultServerSpecification.groovy [hunk 3]
```diff
@@ -146,36 +149,55 @@ class DefaultServerSpecification extends Specification {
...
+                new MongoSocketReadException('Read failed', new ServerAddress(), new IOException()),
+                new MongoSocketReadTimeoutException('Read timed out', new ServerAddress(), new IOException()),
+        ]
     }
 
-    def 'failed open should invalidate the server asychronously'() {
+    def 'failed open should invalidate the server asynchronously'() {
         given:
         def clusterTime = new ClusterClock()
+        def connectionPool = Mock(ConnectionPool)
```
Comment (jyemin) [inline driver-core/src/test/unit/com/mongodb/connection/DefaultServerSpecification.groovy]: No reason not to use a Mock here as well, and then we can check that invalidate is called.

File: driver-core/src/test/unit/com/mongodb/connection/DefaultServerSpecification.groovy (hunk 2 above)
Comment (rozza) [reply driver-core/src/test/unit/com/mongodb/connection/DefaultServerSpecification.groovy]: :)

**Pull request closed without merging**